text/0000-bang-type.md

```diff
@@ -0,0 +1,401 @@
+never contain an element, a `!` enum variant makes the variant guaranteed never
+to occur and so forth. It might seem pointless to use a `!` function argument
+or a `Vec<!>` (just as it would be pointless to use a `()` function argument or
+a `Vec<()>`), but that's no reason to disallow it. And generic code sometimes
+requires it.
+
+Rust already has empty types in the form of empty enums. Any code that could be
+written with this RFC's `!` can already be written by swapping out `!` with
+`Never` (sans implicit casts, see below). So if this RFC could create any
+issues for the language (such as making it unsound or complicating the
+compiler) then these issues would already exist for `Never`.
+
+# Detailed design
+
+Add a type `!` to Rust. `!` behaves like an empty enum except that it can be
+implicitly cast to any other type. ie. the following code is acceptable:
+
+```rust
+let r: Result<i32, !> = Ok(23);
+let i = match r {
+    Ok(i)   => i,
+    Err(e)  => e, // e is cast to i32
+}
+```
+
+Implicit casting is necessary for backwards-compatibility so that code like the
+following will continue to compile:
+
+```rust
+let i: i32 = match some_bool {
+    true  => 23,
+    false => panic!("aaah!"), // an expression of type `!`, gets cast to `i32`
+}
+
+match break {
+    ()  => 23,  // matching with a `()` forces the match argument to be cast to type `()`
+}
+```
+
+In the compiler, remove the distinctions that treat diverging and converging
+expressions as two different kinds of things (eg. stuff like `FnConverging` vs
+`FnDiverging`). Use the type system to do things like reachability analysis.
+
+Add an implementation for `!` of any trait that it can trivially implement. Add
+methods to `Result<T, !>` and `Result<!, E>` for safely extracting the inner
+value. Name these methods along the lines of `unwrap_nopanic`, `safe_unwrap` or
+something.
+
+# Drawbacks
+
+Someone would have to implement this.
+
+# Alternatives
+
+  * Don't do this.
+  * Move @reem's `Void` type into `libcore`. This would create a standard empty
+    type and make it available for use in the standard libraries. If we were to
+    do this it might be an idea to rename `Void` to something else (`Never`,
+    `Empty` and `Mu` have all been suggested). Although `Void` has some
+    precedence in languages like Haskell and Idris the name is likely to trip
+    up people coming from a C/Java et al. background as `Void` is *not* `void`
+    but it can be easy to confuse the two.
+
+# Unresolved questions
+
+Apparently, rust used to have something similar to this but it was removed.
+There are still a few references to `ty_bot` in the compiler. Why was this
+taken out? Note that if there any arguments for not having type `!` in the
+language they should apply equally well to `Never`/`Void` so I assume the old
+`ty_bot` was trying to be something crazier than this RFC's `!` (such as a
+subtype of all types, given the name). Could someone who was around back then
+clarify this?
```

Comment:
> It's not just "slightly more complicated", it's plain impossible due to the information loss.

So then don't lose the information. Keep track of where every `!` came from, whether it's coercible and the AST nodes that would have to be coerced if it is. Here's how I imagine the typechecker working:

```rust
let mut v = if false {
    let mut tmp = Vec::new();
    tmp.push(panic!());
    tmp
} else {
    Vec::new()
};
v.push("foo");
```

The inference variable for the type of `v` is constrained by two things: the `if` expression and the `push("foo")`. First it looks at the `if` expression. The type of the `then` block is the type of `tmp`. The type of `tmp` is constrained by two things: the `Vec::new()` and the `push(panic!())`. `Vec::new()` has type `Vec<! which can be upcast if you coerce AST nodes []>`, `push(panic!())` gives it type `Vec<! which can be upcast if you coerce AST nodes [#12345]>` where #12345 is the node id of `panic!()`. These two types unify to `Vec<! which can be upcast if you coerce AST nodes [#12345]>` giving us the type of `tmp` and a constraint on the type of `v`. Then it looks at the `push("foo")` which constrains the type of `v` to `Vec<&'static str>`. It needs to know the final type for `v` so it unifies `Vec<! which can be upcast if you coerce AST nodes [#12345]>` with `Vec<&'static str>` and gets `Vec<&'static str>`.

By comparison, let's look at something that doesn't typecheck:

```rust
let x = if cond {
    None::<!>
} else {
    None::<&'static str>
}
```

Here, the `if` expression can't be typed because the typechecker can't unify `Option<! which can't be upcast>` with `Option<&'static str>`.